남은 시간 확인 컴포넌트 제작

voidmatcha opened the pull request:
> resolve #14 

- [x] 두자리 유지(1자리의 경우 0 삽입)
- [x] 시간이 1분보다 적어지면 highlight 효과 적용 style, 로직 추가
- [x] Storybook 연결
  - @storybook/addon-controls 사용
- [x] user-select: none; 속성으로 text 선택 못하게 만듬
<img width="167" alt="Screen Shot 2020-10-19 at 2 43 07 AM" src="https://user-images.githubusercontent.com/16266103/96375653-3de36e80-11b5-11eb-81c0-5a3a0420e841.png">
<img width="161" alt="Screen Shot 2020-10-19 at 2 43 14 AM" src="https://user-images.githubusercontent.com/16266103/96375660-4176f580-11b5-11eb-9cc0-5069dbb03c54.png">

Commit:
feat(remain-time): 남은 시간 확인 컴포넌트 제작 (#14)
- [x] 두자리 유지(1자리의 경우 0 삽입)
- [x] 시간이 1분보다 적어지면 highlight 효과 적용 style, 로직 추가
- [x] Storybook 연결
  - @storybook/addon-controls 사용

Lavegaa commented on src/components/RemainTime.js/index.js:
```diff
@@ -0,0 +1,81 @@
+
+RemainTime.defaultProps = {
+  time: 59,
+};
```
file이름을 항상 index로 하시나요? 컴포넌트 이름과 맞췄으면 합니다.
작업 환경에서 index가 여러개 있으면 헛갈릴것 같아요!
index.js에서는 export default만 해주고 실제 컴포넌트 파일은 컴포넌트 명으로 하는건 어떠세요?

voidmatcha replied:
네넵 작업해두신거 참고해서 맞춰놓도록 하겠습니다!

Lavegaa commented on src/components/RemainTime.js/index.js:
```diff
@@ -0,0 +1,81 @@
+  const [second, setSecond] = useState(0);
+  const [toggle, setToggle] = useState(false);
+
+  function numberPad(n, width = 2) {
```
함수는 항상 
```js
function name(arg) {
  return arg;
}
```
와 같은 방법으로 작성하시나요? 이것도 패턴을 맞추면 좋을 것 같아요!
es6문법으로 갈지 기존 방식대로 갈지 의견 주세요!

voidmatcha replied:
arrow function 형태랑 둘 중 뭘 사용할지 정하자는 말씀이신거죠? 제 생각에는 제가 작성한 형태로 맞추는게 더 괜찮을 것 같습니다!

SkynI25 replied:
만약 arrow function과 기본 function 중 정한다면 저도 @dididy 용재님의 의견에 동의합니다!

SkynI25 replied:
음 그렇군요..그냥 function으로 작성하면 this 문제로 이벤트 핸들러로 등록시 `{() => handler()}` 과 같이 작성해줘야 하니까 어떤 방식이 더 나은지 고민되긴 하네요..

SkynI25 replied:
그리고 numberPad 함수는 재사용성이 높아보이는데, utils 함수내 snippet 코드에서 관리하는건 어떨까요? :)

Lavegaa replied:
네 저도 this 문제때문에 함수는 arrow function으로 가는게 좋을것 같은데 기존 방식으로 가는것의 장점이 있을까요?

voidmatcha replied:
@SkynI25 @Lavegaa 
제가 해당 이슈에 대해서 제대로 알고있지 않았네요. arrow function으로 가는 것이 적절해 보입니다.

https://poiemaweb.com/es6-arrow-function#44-addeventlistener-%ED%95%A8%EC%88%98%EC%9D%98-%EC%BD%9C%EB%B0%B1-%ED%95%A8%EC%88%98

앞으로 모든 함수는 arrow function 형태로 작성하는 것으로 하는게 좋을 것 같아요.

Lavegaa commented on package.json:
```diff
@@ -30,6 +30,7 @@
     "@babel/preset-env": "^7.11.5",
     "@babel/preset-react": "^7.10.4",
     "@storybook/addon-actions": "^6.0.26",
+    "@storybook/addon-controls": "^6.0.26",
```
controls를 추가 안해줘도 자동으로 설치 되는것 같아요!

voidmatcha replied:
아하 그렇다면 다음 커밋에는 지워놓도록 하겠습니다!

Commit:
Merge branch 'master' of https://github.com/Yapp-17th/Web_2_Client into feat/#14

SkynI25 commented on src/components/RemainTime.js/index.stories.js:
```diff
@@ -0,0 +1,18 @@
+export const Controls = ({ time = 59 }) => <RemainTime time={time} />;
+
+export const TimeEnough = () => <RemainTime time={200} />;
+export const TimeNotEnough = () => <RemainTime time={40} />;
```
용재님 안녕하세요? 이렇게 `export` 해놓은 변수들은 언제 사용하게 되나요? 초보적인 질문이라 조심스럽지만 스토리북 파일 작성중에 궁금증이 생겨서 여쭤봅니다..!

Lavegaa replied:
스토리북에서 하위 매뉴로 들어가게 됩니다. 
```js
....
const nomal = (args) => <TimeButton {...args} />;
const clicked = (args) => <TimeButton {...args} />;

export const timeButton = nomal.bind({});

export const clickedTimeButton = clicked.bind({});

timeButton.args = { time: 30 };
clickedTimeButton.args = { time: 45 };
....
```
위와 같이 입력하면
![image](https://user-images.githubusercontent.com/34911173/96673385-57490e00-13a1-11eb-987c-ef64850669f7.png)
사진과 같은 이름으로 스토리북에 생깁니다.

voidmatcha replied:
<img width="364" alt="Screen Shot 2020-10-21 at 1 41 56 PM" src="https://user-images.githubusercontent.com/16266103/96674244-50bb9600-13a3-11eb-8727-50e71059eed2.png">

Storybook이 자동으로 export한 변수별로 항목(story)을 만들어줍니다. 

https://storybook.js.org/docs/react/writing-stories/introduction

위 링크에서 잘 설명되어 있습니다!

SkynI25 commented:
전체적으로 깔끔하게 잘 작성하신 듯 합니다..! 코드 작성할 때 많이 참고 하겠습니다 (@dididy @Lavegaa 두분 코드를 통해 많이 배우고 있습니다 ㅎㅎ)

SkynI25 commented on src/components/RemainTime.js/index.js:
```diff
@@ -0,0 +1,81 @@
+  return (
+    <Wrapper highlight={toggle}>
+      <Minutes highlight={toggle}>{minute}</Minutes>
+      <div>:</div>
```
용재님 안녕하세요? 혹시 이 부분을 단순히 `:` 으로 표현하지 않고 div 태그로 감싸신 것은 스타일 적용을 위해서 인거죠? :)

voidmatcha replied:
기태님이 말씀하신 부분을 의도하긴 했는데 사실 이부분은 지워도 상관 없을 것 같네요. 다음 커밋에 반영하겠습니다.

SkynI25 commented:
전체적으로 깔끔한 코드 작성이 인상적이네요!! @dididy  @Lavegaa 두분의 코드를 보면서 많이 배웁니다!!

SkynI25 commented on src/components/RemainTime.js/index.stories.js:
```diff
@@ -0,0 +1,18 @@
+};
+
+// https://storybook.js.org/docs/react/essentials/controls
+export const Controls = ({ time = 59 }) => <RemainTime time={time} />;
```
`Controls` 는 storybook 으로 테스트시 입력값을 받을 수 있도록 하신거군요. 저도 컴포넌트에서 상태값을 관리하게 되면 위와 같이 Controls 를 작성하도록 해야겠습니다!

Commit:
refactor: 코드리뷰 후 수정해야할 사항 적용 (#17)
- 쓸데없는 div 태그 제거
- 컴포넌트 컨벤션에 맞춰 RemainTime.js 파일 추가
- @storybook/addon-controls 삭제
- numberPad 함수 arrow function 형태로 변경(eventHandler this 바인딩
문제)
- default time을 0으로 수정
- numberPad 함수 utils에 추가
- test 코드 작성 및 업데이트

Lavegaa commented on src/components/RemainTime.js/RemainTime.js:
```diff
@@ -0,0 +1,77 @@
+  height: 72px;
+  font-family: TitilliumWebBold;
+  font-size: 25px;
+  color: ${({ highlight }) => (highlight ? '#f2886b' : 'black')};
```
이부분 이제봤네요! props로 하나씩 꺼내서 썻는데 비구조화 할당으로 잘 사용한것 같아서 icon 컴포넌트에 적용했습니다ㅎㅎ 비구조화 할당 잘 쓰기 어려웠는데 하나 배워갑니다!
